reporting/disconnect.html

```diff
@@ -7,19 +7,21 @@
 <script src="/resources/testdriver.js"></script>
 <script src="/resources/testdriver-vendor.js"></script>
 <script>
-  async_test(async function(test) {
-    var observer = new ReportingObserver(function(reports, observer) {
-      test.step(function() {
-        assert_equals(reports.length, 1);
-        assert_equals(reports[0].body.message, "Test message.");
-      });
-      test.done();
+promise_test(async test => {
+    let observer;
+    let observerPromise = new Promise(resolve => {
+      observer = new ReportingObserver(resolve);
+      observer.observe();
+    }).then(reports => {
+      assert_equals(reports.length, 1);
+      assert_equals(reports[0].body.message, "Test message.");
     });
-    observer.observe();
 
     // The observer should still receive this report even though disconnect()
     // is called immediately afterwards.
-    await test_driver.generate_test_report("Test message.");
-    observer.disconnect();
+    let disconnectPromise = test_driver.generate_test_report("Test message.")
+        .then(() => { observer.disconnect(); });
+
+    return Promise.all([observerPromise, disconnectPromise]);
```

Comment:
I'm not sure how you could do this with await. I'd love your help to figure that out. I had real trouble picking the expected flow of this test apart originally, and a `Promise.all` was the only way I could get it to work.

Reply:
I'd ask Paul, but looking at the original it looks to me like the first bit is setting up the stuff that matters to the test, and the test_driver call is just there to generate the reports.

So there is an order dependency, but one has to create the observer first.

Reply:
Right, I see. Ok, I made the ordering explicit, but in doing so I actually discovered something I didn't expect - if the assert fails then I get a *harness level* error.

I need to dig into why, but for now I've uploaded a variant with a deliberately failing error to show this off.

Reply:
Ah, got it. This is a nasty case.

The `ReportingObserver`'s callback is actually called *during* the call to `test_driver.generate_test_report`, not afterwards when we are `await`-ing on `observerPromise`. This places it in a context where testharness.js can't really 'see' it (I'm not very sure why, I find testharness.js very hard to reason about), so the error from the `assert_false` propagates to the unhandledrejection handler and the harness is set to ERROR.

Wrapping the `then` in a `t.step_func` 'works' but is very unintuitive for the reader (imo):

```
diff --git a/reporting/disconnect.html b/reporting/disconnect.html
index ffbfe5a8e9..3907740181 100644
--- a/reporting/disconnect.html
+++ b/reporting/disconnect.html
@@ -12,11 +12,11 @@ promise_test(async test => {
     let observerPromise = new Promise(resolve => {
       observer = new ReportingObserver(resolve);
       observer.observe();
-    }).then(reports => {
+    }).then(test.step_func(reports => {
       assert_false(true, 'Testing what happens if this assert fails')
       assert_equals(reports.length, 1);
       assert_equals(reports[0].body.message, "Test message.");
-    });
+    }));
 
     // The observer should still receive this report even though disconnect()
     // is called immediately afterwards.
```

Reply:
It's roughly equivalent to the following I think.

```
    let firstResolve;
    let promiseA = new Promise(resolve => {
      firstResolve = resolve;
    }).then(reports => {
      assert_false(true, 'Testing what happens if this assert fails')
    });

    await new Promise(resolve => {
      setTimeout(() => {
        firstResolve();
        setTimeout(() => {
          resolve();
        }, 100);
      }, 100);
    });

    await promiseA;
```

I don't have a great solution to this other than to use `t.step_func` and document it clearly. Any thoughts @foolip ?

Reply:
What you ran into in https://github.com/web-platform-tests/wpt/pull/25517/commits/52af27ed382591345a387cad40574087b592b10c was almost certainly an unhandled promise rejection, because nothing was awaiting the result of that promise, the test function was still awaiting the test_driver call.

If you make yourself a `reportsPromise` which can never reject but only resolves with the reports, and put the asserts after `const reports = await reportsPromise`, this problem should go away.

But I admit this is a case where using `Promise.all` makes it obvious that promise rejections at any time are handled.

Reply:
Ah, right, that makes sense.

I've switched to your suggestion; I think it is slightly clearer in terms of the expected ordering (though of course its technically more complex, `observerPromise` will actually be resolved as part of `test_driver.generate_test_report` but we still need to await it to get the results).